File: JavaVersion/AoC2020/src/com/company/Day2.java
package com.company;


import java.util.ArrayList;

public class Day2 extends DefaultClass{

    public Day2()
    {
        super(2,true);
        ArrayList<String> inp = fr.ReadLine();
        ArrayList<String> passwords = new ArrayList<>();
        ArrayList<Integer> MinVal = new ArrayList<>();
        ArrayList<Integer> MaxVal = new ArrayList<>();
        ArrayList<Character> CharVal = new ArrayList<>();
        StringBuilder temp = new StringBuilder();
        for (int i = 0; i < inp.size(); i++) {
            for (int j = 0; j < inp.get(i).length(); j++) {
                if(inp.get(i).charAt(j)=='-') {
                    MinVal.add(Integer.parseInt(temp.toString()));
                    temp = new StringBuilder();
                }
                else if(inp.get(i).charAt(j)==':')
                {
                    MaxVal.add(Integer.parseInt(temp.substring(0,temp.length()-2)));
                    CharVal.add(temp.charAt(temp.length()-1));
                    temp = new StringBuilder();
                }
                else {
                    temp.append(inp.get(i).charAt(j));
                }
            }
            passwords.add(temp.substring(1,temp.length()));
            temp=new StringBuilder();
        }
        int sum=0;
        for (int i = 0; i < inp.size(); i++) {
            if(IsPassword(MinVal.get(i),MaxVal.get(i),CharVal.get(i),passwords.get(i)))
                sum++;
        }
        System.out.println("Part1: "+sum);
        sum=0;
        for (int i = 0; i < inp.size(); i++) {
            if(IsPasswordPart2(MinVal.get(i),MaxVal.get(i),CharVal.get(i),passwords.get(i)))
                sum++;
        }
        System.out.println("Part2: "+sum);
        TimeClock();
    }

    private boolean IsPassword(int min, int max, char C, String password)
    {
        int count=0;
        for (int i = 0; i < password.length(); i++) {
            if(password.charAt(i)==C)
                count++;
        }
        return(count>=min&&count<=max);
    }

    private boolean IsPasswordPart2(int pos1, int pos2, char C, String password)
    {
        if (password.charAt(pos1-1) == C ^ password.charAt(pos2-1) == C)
            return true;
        return false;

    }

}

File: CPPVersion/AoC2020CPP/AoC2020CPP/Day2.cpp
#include "Day2.h"
#include "UsefulMethods.h"

Day2::Day2() :DefaultClass(2, true)
{
	std::vector<std::string> inp = FR.ReadLine();
	values.resize(inp.size());
	std::string temp;
	for (int i = 0; i < inp.size(); i++)
	{
		std::string hold = inp[i];
		for (int x = 0; x < hold.size(); x++)
		{
			if (hold[x] == '-')
			{
				values[i].Min = std::stoi(temp);
				temp = std::string();
			}
			else if (hold[x] == ':')
			{
				values[i].Max = std::stoi(temp.substr(0, temp.size() - 2));
				values[i].C = temp[temp.size() - 1];
				values[i].password = hold.substr(x + 2, hold.size() - 1);
				temp = std::string();
				break;
			}
			else
			{
				temp.append(std::string(1,hold[x]));
			}
		}
	}
	
	int sum=0;
	for (int i = 0; i < inp.size(); i++)
	{
		if (Part1(i))
			sum++;
	}
	std::cout << "Part1: " << sum << "\n";
	sum = 0;
	for (int i = 0; i < inp.size(); i++)
	{
		if (Part2(i))
			sum++;
	}
	std::cout << "Part2: " << sum<< "\n";
	TimeClock();
}

bool Day2::Part1(int n)
{
	int sum=0;
	std::string temp = values[n].password;
	for (int i = 0; i < temp.size(); i++)
	{
		if (temp[i] == values[n].C)
			sum++;
	}
	return(sum > values[n].Min - 1 && sum < values[n].Max + 1);
}

bool Day2::Part2(int i)
{
	return ((values[i].password[values[i].Min - 1] == values[i].C) ^ (values[i].password[values[i].Max - 1] == values[i].C));
}